### ISSUE TITLE
CUB-related test failures

### ISSUE THREAD

**Issue by grlee77** [cupy/cupy#2634]
* Conditions 
CuPy Version          : 7.0.0rc1
CUDA Root             : /usr/local/cuda
CUDA Build Version    : 10010
CUDA Driver Version   : 10010
CUDA Runtime Version  : 10010
cuDNN Build Version   : 7402
cuDNN Version         : 7402
NCCL Build Version    : None
NCCL Runtime Version  : None

* Code to reproduce
run pytest with / without CUB enabled

* Error messages, stack traces, or logs

The following test cases give me an error when running pytest with CUB_DISABALED=0, but pass with CUB_DISABLED=1. I haven't looked into specific cases, but based on the test names I would guess it is a combination of a few possible things:

1.) reduction in the presence of NaN values
2.) reduction of zero-sized arrays
3.) errors in some axis-specific reduction cases

```


 - tests/cupy_tests/core_tests/test_ndarray_reduction.py:39 TestArrayReduction.test_max_nan
 - tests/cupy_tests/core_tests/test_ndarray_reduction.py:51 TestArrayReduction.test_max_nan_imag
 - tests/cupy_tests/core_tests/test_ndarray_reduction.py:45 TestArrayReduction.test_max_nan_real
 - tests/cupy_tests/core_tests/test_ndarray_reduction.py:87 TestArrayReduction.test_min_nan
 - tests/cupy_tests/core_tests/test_ndarray_reduction.py:99 TestArrayReduction.test_min_nan_imag
 - tests/cupy_tests/core_tests/test_ndarray_reduction.py:93 TestArrayReduction.test_min_nan_real
 - tests/cupy_tests/linalg_tests/test_einsum.py:279 TestEinSumUnaryOperation_param_20.test_einsum_unary
 - tests/cupy_tests/linalg_tests/test_einsum.py:297 TestEinSumUnaryOperation_param_20.test_einsum_unary_dtype
 - tests/cupy_tests/linalg_tests/test_einsum.py:279 TestEinSumUnaryOperation_param_49.test_einsum_unary
 - tests/cupy_tests/linalg_tests/test_einsum.py:297 TestEinSumUnaryOperation_param_49.test_einsum_unary_dtype
 - tests/cupy_tests/linalg_tests/test_einsum.py:279 TestEinSumUnaryOperation_param_136.test_einsum_unary
 - tests/cupy_tests/linalg_tests/test_einsum.py:297 TestEinSumUnaryOperation_param_136.test_einsum_unary_dtype
 - tests/cupy_tests/linalg_tests/test_einsum.py:289 TestEinSumUnaryOperation_param_136.test_einsum_unary_views
 - tests/cupy_tests/linalg_tests/test_norms.py:18 TestTrace.test_external_trace
 - tests/cupy_tests/linalg_tests/test_norms.py:12 TestTrace.test_trace
 - tests/cupy_tests/padding_tests/test_pad.py:21 TestPadDefault_param_9.test_pad_default
 - tests/cupy_tests/padding_tests/test_pad.py:21 TestPadDefault_param_10.test_pad_default
 - tests/cupy_tests/padding_tests/test_pad.py:21 TestPadDefault_param_11.test_pad_default
 - tests/cupy_tests/padding_tests/test_pad.py:21 TestPadDefault_param_12.test_pad_default
 - tests/cupy_tests/padding_tests/test_pad.py:21 TestPadDefault_param_13.test_pad_default
 - tests/cupy_tests/padding_tests/test_pad.py:21 TestPadDefault_param_14.test_pad_default
 - tests/cupy_tests/padding_tests/test_pad.py:133 TestPad_param_9.test_pad
 - tests/cupy_tests/padding_tests/test_pad.py:133 TestPad_param_10.test_pad
 - tests/cupy_tests/padding_tests/test_pad.py:133 TestPad_param_11.test_pad
 - tests/cupy_tests/padding_tests/test_pad.py:133 TestPad_param_12.test_pad
 - tests/cupy_tests/padding_tests/test_pad.py:133 TestPad_param_13.test_pad
 - tests/cupy_tests/padding_tests/test_pad.py:133 TestPad_param_14.test_pad
 - tests/cupy_tests/padding_tests/test_pad.py:133 TestPad_param_15.test_pad
 - tests/cupy_tests/padding_tests/test_pad.py:133 TestPad_param_16.test_pad
 - tests/cupy_tests/sorting_tests/test_search.py:23 TestSearch.test_argmax_nan
 - tests/cupy_tests/sorting_tests/test_search.py:65 TestSearch.test_argmax_zero_size
 - tests/cupy_tests/sorting_tests/test_search.py:89 TestSearch.test_argmin_nan
 - tests/cupy_tests/sorting_tests/test_search.py:137 TestSearch.test_argmin_zero_size
 - tests/cupy_tests/sorting_tests/test_search.py:327 TestNanArgMin.test_nanargmin_zero_size
 - tests/cupy_tests/sorting_tests/test_search.py:417 TestNanArgMax.test_nanargmax_zero_size
```

cc @leofang and @anaruse as fellow CUB contributors here

ref: #2579 which should help avoid similar issues in the future

**Comment by leofang** [comment]
Thanks for the thorough tests, @grlee77! Apparently these corner cases are not captured in #2598 either... 😞 
> 1.) reduction in the presence of NaN values

This is super tricky. If I am not wrong, the problem is in CUB: in [this file](https://github.com/NVlabs/cub/blob/1.8.0/cub/thread/thread_operators.cuh) the comparison operators do not even bother to check `isnan()`. (Note that `CUB_MAX`/`CUB_MIN` are simple macros defined [here](https://github.com/NVlabs/cub/blob/1.8.0/cub/util_macro.cuh).)

If we bundle the CUB headers as a submodule, we won't be able to tweak the CUB source code, and have to address this at the Cython level.

If we keep a copy of CUB headers (as was originally done in #2584), we may be able to test `isnan()` in the operators mentioned above. 

I need input from @kmaehashi before preceding to test this. 

> 2.) reduction of zero-sized arrays

This is easy to solve. Just test `input_array.size` and raise wherever necessary.

> 3.) errors in some axis-specific reduction cases

This is also easy. `device_segmented_reduce()` assumed a contiguous input, but often with `einsum()` this is not guaranteed. In this case, we just fallback to the old reduction kernel.

I have not finished examining all the errors, so we may find more bugs. Working on it...

**Comment by leofang** [comment]
(Updated my reply above.)

**Comment by kmaehashi** [comment]
Hmm, I'm trying to reproduce but currently unable to build CuPy master with CUB 1.8.0, maybe I'm missing something...

`cupy/core/include/cupy/complex/arithmetic.h:262:80: error: ‘double thrust::abs(const thrust::complex<double>&)’ is not declared in ‘thrust’`

<details>

```
building 'cupy.cuda.cub' extension
gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -D_FORCE_INLINES=1 -I/usr/local/cuda/include -I/home/kenichi/Development/cupy/install/../cupy/core/include -I/home/kenichi/Development/cub -I/home/kenichi/.pyenv/versions/3.7.0/envs/local-3.7.0/include -I/home/kenichi/.pyenv/versions/3.7.0/include/python3.7m -c cupy/cuda/cub.cpp -o build/temp.linux-x86_64-3.7/cupy/cuda/cub.o -fopenmp
    NVCC options: ['--generate-code=arch=compute_52,code=sm_52', '-O2', '--compiler-options="-fPIC"', '--std=c++11']
/usr/local/cuda/bin/nvcc -D_FORCE_INLINES=1 -I/usr/local/cuda/include -I/home/kenichi/Development/cupy/install/../cupy/core/include -I/home/kenichi/Development/cub -I/home/kenichi/.pyenv/versions/3.7.0/envs/local-3.7.0/include -I/home/kenichi/.pyenv/versions/3.7.0/include/python3.7m -c cupy/cuda/cupy_cub.cu -o build/temp.linux-x86_64-3.7/cupy/cuda/cupy_cub.o --generate-code=arch=compute_52,code=sm_52 -O2 --compiler-options="-fPIC" --std=c++11                                                                                                                            /home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/arithmetic.h:258:78: error: ‘float thrust::abs(const thrust::complex<float>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/arithmetic.h:262:80: error: ‘double thrust::abs(const thrust::complex<double>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/cexp.h:169:91: error: ‘thrust::complex<double> thrust::exp(const thrust::complex<double>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/cexpf.h:149:89: error: ‘thrust::complex<float> thrust::exp(const thrust::complex<float>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/clog.h:192:91: error: ‘thrust::complex<double> thrust::log(const thrust::complex<double>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/clogf.h:188:89: error: ‘thrust::complex<float> thrust::log(const thrust::complex<float>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/ccosh.h:195:2: error: ‘thrust::complex<double> thrust::cos(const thrust::complex<double>&)’ is not declared in ‘thrust’
     const thrust::complex<double>& z) {
  ^
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/ccosh.h:201:2: error: ‘thrust::complex<double> thrust::cosh(const thrust::complex<double>&)’ is not declared in ‘thrust’
     const thrust::complex<double>& z) {
  ^
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/ccoshf.h:126:89: error: ‘thrust::complex<float> thrust::cos(const thrust::complex<float>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/ccoshf.h:131:90: error: ‘thrust::complex<float> thrust::cosh(const thrust::complex<float>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/csinh.h:183:91: error: ‘thrust::complex<double> thrust::sin(const thrust::complex<double>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/csinh.h:188:92: error: ‘thrust::complex<double> thrust::sinh(const thrust::complex<double>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/csinhf.h:124:89: error: ‘thrust::complex<float> thrust::sin(const thrust::complex<float>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/csinhf.h:129:90: error: ‘thrust::complex<float> thrust::sinh(const thrust::complex<float>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/ctanh.h:182:91: error: ‘thrust::complex<double> thrust::tan(const thrust::complex<double>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/ctanhf.h:107:89: error: ‘thrust::complex<float> thrust::tan(const thrust::complex<float>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/csqrt.h:140:92: error: ‘thrust::complex<double> thrust::sqrt(const thrust::complex<double>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/csqrtf.h:137:90: error: ‘thrust::complex<float> thrust::sqrt(const thrust::complex<float>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/catrig.h:697:92: error: ‘thrust::complex<double> thrust::acos(const thrust::complex<double>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/catrig.h:702:92: error: ‘thrust::complex<double> thrust::asin(const thrust::complex<double>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/catrig.h:708:92: error: ‘thrust::complex<double> thrust::atan(const thrust::complex<double>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/catrig.h:725:93: error: ‘thrust::complex<double> thrust::atanh(const thrust::complex<double>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/catrigf.h:411:90: error: ‘thrust::complex<float> thrust::acos(const thrust::complex<float>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/catrigf.h:416:90: error: ‘thrust::complex<float> thrust::asin(const thrust::complex<float>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/catrigf.h:422:90: error: ‘thrust::complex<float> thrust::atan(const thrust::complex<float>&)’ is not declared in ‘thrust’
/home/kenichi/Development/cupy/install/../cupy/core/include/cupy/complex/catrigf.h:439:91: error: ‘thrust::complex<float> thrust::atanh(const thrust::complex<float>&)’ is not declared in ‘thrust’
error: command '/usr/local/cuda/bin/nvcc' failed with exit status 1
```

**Comment by leofang** [comment]
@kmaehashi Your compiler is probably too old. If using gcc, it seems the 4.x series is not fully supporting c++11, despite they claimed so. I've had a hard time with it, see https://github.com/cupy/cupy/issues/2530#issuecomment-539366174...

**Comment by kmaehashi** [comment]
Thanks @leofang, I'm currently on Ubuntu 16 (gcc 5.4.0 / nvcc 9.2), so I guess it's enough (and it can also build `cupy.cuda.thrust`). Will try on another environment.

**Comment by leofang** [comment]
That's odd. I think I tried this combination earlier and it seemed to work. Will look into this later.

**Comment by leofang** [comment]
Found a way to treat NaNs without touching CUB source code! For example, this works for fixing `Max()`:
```diff
diff --git a/cupy/cuda/cupy_cub.cu b/cupy/cuda/cupy_cub.cu
index ad78a89..cd73383 100644
--- a/cupy/cuda/cupy_cub.cu
+++ b/cupy/cuda/cupy_cub.cu
@@ -6,6 +6,31 @@
 
 using namespace cub;
 
+// 1. The function is appended "cupy_" to avoid confusion and name collision
+// 2. std::isnan() cannot be used here as <cmath> cannot be included
+template <typename T>
+__host__ __device__ __forceinline__ bool cupy_isnan(const T &a) { return a != a; }
+
+// specialization for float for handling NaNs
+// NumPy behavior: NaN is always chosen!
+template <>
+__host__ __device__ __forceinline__ float Max::operator()(const float &a, const float &b) const
+{
+    if (cupy_isnan(a)) {return a;}
+    else if (cupy_isnan(b)) {return b;}
+    else {return CUB_MAX(a, b);}
+}
+
+// specialization for double for handling NaNs
+// NumPy behavior: NaN is always chosen!
+template <>
+__host__ __device__ __forceinline__ double Max::operator()(const double &a, const double &b) const
+{
+    if (cupy_isnan(a)) {return a;}
+    else if (cupy_isnan(b)) {return b;}
+    else {return CUB_MAX(a, b);}
+}
+
 // Minimum boilerplate to support complex numbers in sum(), min(), and max():
 // - This works only because all data fields in the *Traits struct are not
 //   used in <cub/device/device_reduce.cuh>.
```
Will add and expand this to #2636 later today.

**Comment by leofang** [comment]
> `// NumPy behavior: NaN is always chosen!`

@kmaehashi, given that, I don't think these lines are correct:
https://github.com/cupy/cupy/blob/fff3d28c2d3dbfb95254a9ea860d55a54a3b7e4c/cupy/core/include/cupy/complex.cuh#L53-L56
https://github.com/cupy/cupy/blob/fff3d28c2d3dbfb95254a9ea860d55a54a3b7e4c/cupy/core/include/cupy/complex.cuh#L68-L71
Can I correct them in #2636 as well?

**Comment by grlee77** [comment]
Great work on finding this workaround. 

I can confirm that the other NaN case you find is also a bug. For
```Python
xp.max(xp.array([5, -1, np.nan]))
```
Numpy gives NaN, but CuPy gives 5. The CuPy case is what you would want if you are trying to find the max while ignoring NaNs.

Basically `cupy.max` is behaving like `numpy.nanmax`.

**Comment by leofang** [comment]
> Can I correct them in #2636 as well?

Those lines do not seem to have actual impact at the Python level. It's likely that the reduction kernel implementation took care of it.

> Basically `cupy.max` is behaving like `numpy.nanmax`.

I hope all of these nuisances are addressed in #2636 now. PTAL.

**Comment by kmaehashi** [comment]
> Hmm, I'm trying to reproduce but currently unable to build CuPy master with CUB 1.8.0, maybe I'm missing something...
> cupy/core/include/cupy/complex/arithmetic.h:262:80: error: ‘double thrust::abs(const thrust::complex<double>&)’ is not declared in ‘thrust’

For record: Using `gcc-7` instead of `gcc` (v5.4.0) by specifying `--compiler-bindir gcc-7` to `nvcc` (`export NVCC="ccache nvcc --compiler-bindir gcc-7"`) solved the issue on Ubuntu 16.04.

cc/ @anaruse

**Comment by leofang** [comment]
FYI, with gcc6+ from conda this is what I usually do: `export NVCC='nvcc -ccbin /path/to/conda/environment/bin/x86_64-conda_cos6-linux-gnu-gcc'`. 

@kmaehashi @anaruse
